armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+                )
+            else:
+                meters_args.append((meter, arg))
+
+    def disconnect_meter(self, meter):
+        """
+        Disconnect meter from probes; idempotent
+        """
+        for arg in meter.get_arg_names():
+            probe_variable, stage_filter = process_meter_arg(arg)
+            if probe_variable not in self.probe_filter_meter_arg:
+                continue
+            filter_map = self.probe_filter_meter_arg[probe_variable]
+
+            if stage_filter not in filter_map:
+                continue
+            meters_args = filter_map[stage_filter]
+
+            if (meter, arg) in meters_args:
+                meters_args.remove((meter, arg))
+                if not meters_args:
+                    filter_map.pop(stage_filter)
+                if not filter_map:
+                    self.probe_filter_meter_arg.pop(probe_variable)
+
+    def map_to_meters_args(self, probe_variable, stage):
+        """
+        Return a list of (meter, arg) that are using the current probe_variable
+        """
+        filter_map = self.probe_filter_meter_arg.get(probe_variable, {})
+        meters = filter_map.get(stage, [])
+        meters.extend(filter_map.get(None, []))  # no stage filter (default)
+        return meters
+
+
+class Context:  # NOTE: may need to rename to something like Experiment or Procedure
+    def __init__(self, name="global"):
+        self.name = name
+        self.batch = -1
+        self.stage = ""
+        self.mapper = ProbeMapper()
+        self.meters = []
+        self.writers = []
+        self.closed = False
+
+    def set_stage(self, stage: str):
+        self.stage = stage
+
+    def set_batch(self, batch: int):
+        # NOTE: batch could be a set of sample indices
+        self.batch = batch
+
+    def increment_batch(self):
+        self.batch += 1
+
+    def is_measuring(self, probe_variable):
+        return bool(self.mapper.map_to_meters_args(probe_variable, self.stage))
+
+    def update(self, probe_variable, value):
+        meters_args = self.mapper.map_to_meters_args(probe_variable, self.stage)
+        if not meters_args:
+            raise ValueError("No meters are measuring")
+        for meter, arg in meters_args:
+            meter.set(arg, value, self.batch)
+
+    def connect_meter(self, meter):
+        if meter in self.meters:
+            return
+
+        self.meters.append(meter)
+        self.mapper.connect_meter(meter)
+        for writer in self.writers:
+            meter.add_writer(writer)
+
+    def get_meters(self):
+        return self.meters
+
+    def add_writer(self, writer):
+        """
+        Convenience method to add writer to all meters in this context
+        """
+        if writer in self.writers:
+            log.warning(f"writer {writer} already connected to {self.name} context")
+            return
+
+        self.writers.append(writer)
+        for meter in self.meters:
```

Comment:
So this current approach adds the writer to _all_ meters. What if, for example, one thing I want to do is write a JPG file of an image after going through a model's preprocessing ops. I imagine I'd need to create a `Writer` class that does what I want with the image. If this is the case, I wouldn't want this writer to apply to other meters. Should a writer thus be tied to a specific meter then?

Reply:
Yes, you would want to create a `Writer` and apply it to that specific Meter. Need to think about how to have optional pieces here.